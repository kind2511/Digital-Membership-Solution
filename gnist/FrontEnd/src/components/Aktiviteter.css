.aktiviteter-container-unique {
  display: flex;
  flex-direction: column;
}

.section-unique {
  padding: 22px;
  border-radius: 5px;
  box-shadow: 0 2px 6px rgba(0, 0, 0, 0.1);
  background-color: #fff;
}

.section-title-unique {
  margin-bottom: 20px;
  color: #333;
  font-size: 24px;
}

.activity-form-unique input[type="text"],
.activity-form-unique input[type="date"],
.activity-form-unique textarea {
  width: 100%;
  padding: 5px;
  margin-bottom: 20px;
  border-radius: 5px;
  border: 1px solid #ccc;
}

.activity-form-unique input[type="file"] {
  margin-bottom: 15px;
}

.activity-form-unique button {
  padding: 10px 20px;
  border: none;
  background-color: #005A9E;
  color: white;
  cursor: pointer;
  border-radius: 5px;
}

.activity-form-unique button:hover {
  background-color: #004080;
}

.activity-success-message-unique {
  transition: opacity 0.5s ease-in-out;
  color: #FFFFFF;
  text-align: center;
  border-radius: 5px;
  padding: 10px;
  margin-bottom: 20px;
  background-color: #52C41A;
  overflow: hidden;
  white-space: nowrap;
}

.activities-list-unique {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
  gap: 20px;
}

.activity-item-unique {
  display: flex;
  flex-direction: column;
  gap: 10px;
  padding: 15px;
  background-color: rgba(0, 0, 0, 0.1);
  border-radius: 10px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  transition: transform 0.2s ease-in-out, box-shadow 0.2s ease-in-out; 
  cursor: pointer;
}

.activity-item-unique:hover {
  transform: translateY(-5px);
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2);
}

.activity-item-unique img {
  width: 100%; 
  height: 200px;  /* Set a fixed height */
  object-fit: cover; 
  border-radius: 5px; 
}


.activity-info-unique h3 {
  margin: 0;
  color: #005A9E;
  font-size: 18px;
}

.activity-date-unique {
  font-size: 14px;
  color: #666;
}

.activity-description-unique {
  font-size: 14px;
  color: #333;
}

.alle-aktiviteter-unique {
  margin-top: 20px;
}

/* Modal Styles */
.modal-unique {
  position: fixed;
  z-index: 1000;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  justify-content: center;
  align-items: center;
}

.modal-content-unique {
  background-color: #fefefe;
  padding: 20px;
  border-radius: 5px;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
  position: relative;
  width: 500px;
  height: 500px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.close-button-unique {
  background-color: #D50000;
  color: white;
  padding: 10px 20px;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  position: absolute;
  top: 10px;
  right: 10px;
}

.registrants-list-unique {
  list-style: decimal inside; 
  padding: 0; 
  width: 100%;
  text-align: left;
  max-height: 400px;
  overflow-y: auto;
}

.registrant-unique {
  margin: 5px 0; 
  font-size: 16px; 
  line-height: 1.5; 
}

.modal-footer-unique {
  display: flex;
  justify-content: flex-end;
  padding-top: 20px;
  position: absolute;
  bottom: 10px;
  right: 10px;
}

.delete-button-unique {
  background-color: #D50000;
  color: white;
  padding: 10px 8px;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  margin-right: 100px; /* space between the buttons */
  margin-top: -10px;
}

.close-button-unique {
  background-color: #FFA500;
  color: white;
  padding: 10px 20px;
  border: none;
  border-radius: 5px;
  cursor: pointer;

}

.delete-button-unique:hover{
  background-color: darkred;
}

.close-button-unique:hover{
  background-color: darkorange;
}


/* modal width on smaller screens */
@media (max-width: 768px) {
  .modal-content-unique {
    width: 80%;
    height: 80%;
  }

  .aktiviteter-container-unique {
    flex-direction: column;
  }

  .section-unique {
    margin-bottom: 20px;
  }

  .activity-item-unique img {
    width: 100%;
    height: auto;
  }
}
