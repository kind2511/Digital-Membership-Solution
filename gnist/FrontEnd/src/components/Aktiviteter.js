import React, { useState, useEffect } from 'react';
import axios from 'axios';
import './Aktiviteter.css';

function Aktiviteter() {
  const [activities, setActivities] = useState([]);
  const [registrants, setRegistrants] = useState([]);
  const [activity, setActivity] = useState({
    dato: '',
    tittel: '',
    bilde: null,
    beskrivelse: '',
    limit: '',  
  });
  const [successMessage, setSuccessMessage] = useState('');
  const [showSuccess, setShowSuccess] = useState(false);
  const [selectedActivity, setSelectedActivity] = useState(null);
  const [archivedActivities, setArchivedActivities] = useState([]);

  useEffect(() => {
    const fetchActivities = async () => {
      try {
        const response = await axios.get('http://127.0.0.1:8000/digital_medlemsordning/get_future_activities/');
        setActivities(response.data);
      } catch (error) {
        console.error("Error fetching activities:", error);
      }
    };

    fetchActivities();
  }, []);

  const handleChange = (e) => {
    const { name, value, files } = e.target;
    setActivity(prev => ({
      ...prev,
      [name]: name === 'bilde' ? files[0] : value
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();

    const formData = new FormData();
    formData.append('image', activity.bilde);
    formData.append('title', activity.tittel);
    formData.append('description', activity.beskrivelse);
    formData.append('date', activity.dato);
    formData.append('limit', activity.limit);

    try {
      await axios.post('http://127.0.0.1:8000/digital_medlemsordning/create_activity/', formData, {
        headers: {
          'Content-Type': 'multipart/form-data'
        }
      });
      setSuccessMessage('Aktivitetet ble lagret');
      setShowSuccess(true);
      setTimeout(() => {
        setShowSuccess(false);
        setActivity({
          dato: '',
          tittel: '',
          bilde: null,
          beskrivelse: '',
          limit: '',
        });
      }, 5000);

    } catch (error) {
      console.error("Error creating activity:", error);
    }
  };

  const fetchRegistrants = async (activityId) => {
    try {
      const response = await axios.get(`http://127.0.0.1:8000/digital_medlemsordning/get_signed_up_members/${activityId}/`);
      setRegistrants(response.data.sign_up_members);
      setSelectedActivity(activityId);
    } catch (error) {
      console.error("Error fetching registrants:", error);
    }
  };

  const handleDeleteActivity = async (activityId) => {
    try {
      const response = await axios.delete(`http://127.0.0.1:8000/digital_medlemsordning/delete_activity/${activityId}/`);
      console.log(response.data.message);
      setActivities(currentActivities => currentActivities.filter(act => act.activityID !== activityId));
      setSelectedActivity(null);
      setSuccessMessage('Aktivitet slettet');
      setShowSuccess(true);
      setTimeout(() => {
        setShowSuccess(false);
      }, 3000);

    } catch (error) {
      console.error("Error deleting activity:", error);
    }
  };

  useEffect(() => {
    const fetchArchivedActivities = async () => {
      try {
        const response = await axios.get('http://127.0.0.1:8000/digital_medlemsordning/get_past_activities/');
        setArchivedActivities(response.data);
      } catch (error) {
        console.error("Error fetching archived activities:", error);
      }
    };

    fetchArchivedActivities();
  }, []);

  return (
    <div className="aktiviteter-container-unique">
      <div className="activity-success-message-unique" style={{ opacity: showSuccess ? 1 : 0, height: showSuccess ? 'auto' : '0' }}>
        {successMessage}
      </div>

      {/* Modal for Registrants */}
      {selectedActivity && (
        <div className="modal-unique">
          <div className="modal-content-unique">
            <h3>Folk som har registrert seg p√• denne aktiviteten:</h3>
            <ol className="registrants-list-unique">
              {registrants.map((person, index) => (
                <li key={index} className="registrant-unique">
                  {person.first_name} {person.last_name}
                </li>
              ))}
            </ol>
            <div className="modal-footer-unique">
              <button className="delete-button-unique" onClick={() => handleDeleteActivity(selectedActivity)}>Slett Aktivitet</button>
              <button className="close-button-unique" onClick={() => setSelectedActivity(null)}>Lukk</button>
            </div>
          </div>
        </div>
      )}

      {/* Lag Ny Aktivitet Section */}
      <div className="section-unique lag-ny-aktivitet-unique">
        <h2 className="section-title-unique">Lag Ny Aktivitet</h2>
        <form onSubmit={handleSubmit} className="activity-form-unique">
          <input type="date" name="dato" value={activity.dato} onChange={handleChange} required />
          <input type="text" name="tittel" placeholder="Tittel" value={activity.tittel} onChange={handleChange} required />
          <input type="file" name="bilde" onChange={handleChange} />
          <textarea name="beskrivelse" placeholder="Beskrivelse" value={activity.beskrivelse} onChange={handleChange} required />
          <input type="number" name="limit" placeholder="Antall Plasser" value={activity.limit} onChange={handleChange} required />
          <button type="submit">Lagre Aktivitet</button>
        </form>
      </div>

      {/* Kommende Aktiviteter Section */}
      <div className="section-unique alle-aktiviteter-unique">
        <h2 className="section-title-unique">Kommende Aktiviteter</h2>
        <div className="activities-list-unique">
          {activities.map((activity) => (
            <div className="activity-item-unique" key={activity.activityID} onClick={() => fetchRegistrants(activity.activityID)}>
              <img src={activity.image} alt={activity.title} />
              <div className="activity-info-unique">
                <h3>{activity.title}</h3>
                <p className="activity-date-unique">{activity.date}</p>
                <p className="activity-description-unique">{activity.description}</p>
                <p className="activity-places-unique">Antall Plasser: {activity.limit}</p>
              </div>
            </div>
          ))}
        </div>
      </div>

      {/* Arkiverte Aktiviteter Section */}
      <div className="section-unique alle-aktiviteter-unique">
        <h2 className="section-title-unique">Arkiverte Aktiviteter</h2>
        <div className="activities-list-unique">
          {archivedActivities.map((archivedActivity) => (
            <div className="activity-item-unique" key={archivedActivity.activityID} onClick={() => fetchRegistrants(archivedActivity.activityID)}>
              <img src={archivedActivity.image} alt={archivedActivity.title} />
              <div className="activity-info-unique">
                <h3>{archivedActivity.title}</h3>
                <p className="activity-date-unique">{archivedActivity.date}</p>
                <p className="activity-description-unique">{archivedActivity.description}</p>
              </div>
            </div>
          ))}
        </div>
      </div>
    </div>
  );

}

export default Aktiviteter;
