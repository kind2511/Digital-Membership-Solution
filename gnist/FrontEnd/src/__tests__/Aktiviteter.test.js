import React from 'react';
import renderer from 'react-test-renderer';
import Aktiviteter from '../components/Aktiviteter';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

// Mock axios for the test
const mock = new MockAdapter(axios);
mock.onGet('http://127.0.0.1:8000/digital_medlemsordning/get_future_activities/').reply(200, []);
mock.onGet('http://127.0.0.1:8000/digital_medlemsordning/get_past_activities/').reply(200, []);

test('Aktiviteter component snapshot', () => {
  const component = renderer.create(<Aktiviteter />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
