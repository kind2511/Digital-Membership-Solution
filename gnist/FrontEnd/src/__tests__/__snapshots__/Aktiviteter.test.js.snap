// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aktiviteter component snapshot 1`] = `
<div
  className="aktiviteter-container-unique"
>
  <div
    className="activity-success-message-unique"
    style={
      Object {
        "height": "0",
        "opacity": 0,
      }
    }
  />
  <div
    className="section-unique lag-ny-aktivitet-unique"
  >
    <h2
      className="section-title-unique"
    >
      Lag Ny Aktivitet
    </h2>
    <form
      className="activity-form-unique"
      onSubmit={[Function]}
    >
      <input
        name="dato"
        onChange={[Function]}
        required={true}
        type="date"
        value=""
      />
      <input
        name="tittel"
        onChange={[Function]}
        placeholder="Tittel"
        required={true}
        type="text"
        value=""
      />
      <input
        name="bilde"
        onChange={[Function]}
        type="file"
      />
      <textarea
        name="beskrivelse"
        onChange={[Function]}
        placeholder="Beskrivelse"
        required={true}
        value=""
      />
      <input
        name="limit"
        onChange={[Function]}
        placeholder="Antall Plasser"
        type="number"
        value=""
      />
      <button
        type="submit"
      >
        Lagre Aktivitet
      </button>
    </form>
  </div>
  <div
    className="section-unique alle-aktiviteter-unique"
  >
    <h2
      className="section-title-unique"
    >
      Kommende Aktiviteter
    </h2>
    <div
      className="activities-list-unique"
    />
  </div>
  <div
    className="section-unique alle-aktiviteter-unique"
  >
    <h2
      className="section-title-unique"
    >
      Arkiverte Aktiviteter
    </h2>
    <div
      className="activities-list-unique"
    />
  </div>
</div>
`;
